### ISSUE TITLE
Custom Getter

### ISSUE THREAD

**Issue by A-312** [mozilla/node-convict#313]
We should allow to set custom getter to add more choice between `env`, `arg` and `.load(`.

## Example: Custom Getter for file

```js
const schema = {
  db: {
    name: {
      doc: 'Database name',
      format: String,
      default: 'users',
      getter: 'file',
      filepath: '/mycustompath/togetmy/name'
    }
  }
}

// convict.addGetter(name, priority, getter)
convict.addGetter('file', 70, (name, val, schema) => readFile(schema.filepath))
```

## Custom priority level

And we should allow to set custom priority level : https://github.com/mozilla/node-convict/blob/master/lib/convict.js#L600-L601

| Priority level | Getter |
| - | - |
| 0 | Default value |
| 20 | File or json set in function argument |
| 40 | Environment variables (only used when env property is set in schema) |
| 60 | Command line arguments (only used when arg property is set in schema) |

**Comment by A-312** [comment]
I think there is two types possible for Getter : Loading only once like (env, arg) or loading for each `.load` like this : 

And priority argument can be number (=> getter is loaded only once) or string ('load' or 'load/set') (=> getter is loaded for each .load (if the value is set)).

```js
const schema = {
  options: {
    format: Array,
    default: [],
    getter: 'array'
  }
}

// convict.addGetter(name, priority, getter)
// priority number or string ('load' | 'load/set')
convict.addGetter('array', 'load/set', (name, val, schema, current) => {
  if (Array.isArray(val))
    return val;
  elseif (typeof val === 'object' && Array.isArray(val))
    val.add.forEach((n) => current.push(n))
  else
    throw 'invalid value for Array getter'
})

convict.load({
  options: ['utf8']
})

convict.load({
  options: { add: ['table', 'verbose'] }
})

convict.validate();

// convict.get(options) // ['utf8', 'table', 'verbose']
```

what did you think ?

**Comment by A-312** [comment]
Custom Getter can be used to make a support for #322 -> formats: ["ip", "port"]

**Comment by A-312** [comment]
I will change the logic of : 

https://github.com/mozilla/node-convict/blob/be350a7ded016543b3eea16953be1f81c4bf34be/lib/convict.js#L329-L353

https://github.com/mozilla/node-convict/blob/be350a7ded016543b3eea16953be1f81c4bf34be/lib/convict.js#L250-L264

**Comment by A-312** [comment]
I will make default, env and argv like a customGetter.

**Comment by A-312** [comment]
Instead todo (6 loops) : 

(Order : default, value *(=.load or .loadFile)*, env, arg, set *(= .set)*)

// loading schema

  1. default loop
  2. env loop
  3. arg loop

// .load or .loadFile

  4. value loop
  5. env loop
  6. arg loop

I will do (2 loops) : 

// loading schema

   1. Check for the value with the most priority --> 1) arg, 2) env, 3) default and get the value. I will write a variable _cvtActualGetter in schema (by the getter name arg/env/default.

// .load or .loadFile

   2. Check the value with less priority of 'value' in `_cvtActualGetter ` to run 'default'.

**Comment by A-312** [comment]
I referenced some issue to fix it in my PR.

**Comment by A-312** [comment]
The actual prototype is : addGetter(property, getter, useCoerce, usedOnlyOnce, rewrite)

 - property : Name & keyname in schema
 - getter : function to get the value
 - useCoerce : true/false to apply coerce (will fix: #68)
 - usedOnlyOnce : boolean or function, if true will fire : `throw new Error("'" + fullName + "' reuses '" + getterName + "' getter value: " + value + advice);` ;
 - rewrite : If true you can rewrite a getter with the same name (property).
